### PR TITLE
Closes #2259 - Refactor UpdateTaskAccTest to use test-api

### LINKED ISSUE #2259: Refactor UpdateTaskAccTest to use test-api

### PR DESCRIPTION
**Pull request by jamesrdi**
https://sonarcloud.io/summary/new_code?id=jamesrdi_taskana&branch=TSK-2259

---
Release Notes:

```

```
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [ ] After integration of the PR, I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [ ] Commit message format → TSK-XXX: Your commit message.
- [ ] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [ ] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+  @WithAccessId(user = "user-1-2")
+  @Test
+  void should_ThrowException_When_TaskHasAlreadyBeenUpdated() throws Exception {
+    Task task =
+        TaskBuilder.newTask()
+            .classificationSummary(defaultClassificationSummary)
+            .workbasketSummary(defaultWorkbasketSummary)
+            .primaryObjRef(defaultObjectReference)
+            .buildAndStore(taskService, "admin");
+    final Task task2 = taskService.getTask(task.getId());
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:234]: You don't need final here. Please use copy() instead of getTask

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+        defaultTestClassification().buildAndStoreAsSummary(classificationService, "admin");
+    task.setClassificationKey(classificationSummary.getKey());
+    Task updatedTask = taskService.updateTask(task);
+    updatedTask = taskService.getTask(updatedTask.getId());
+
+    assertThat(updatedTask).isNotNull();
+    assertThat(updatedTask.getClassificationSummary().getKey())
+        .isEqualTo(classificationSummary.getKey());
+    assertThat(updatedTask.getClassificationSummary()).isNotEqualTo(defaultClassificationSummary);
+    assertThat(updatedTask.getCreated()).isNotEqualTo(updatedTask.getModified());
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:287]: You don't need to check "modified" here

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+  @WithAccessId(user = "user-1-2")
+  @Test
+  void should_UpdateReadFlagOfTask_When_SetTaskReadIsCalled() throws Exception {
+    Task task =
+        TaskBuilder.newTask()
+            .classificationSummary(defaultClassificationSummary)
+            .workbasketSummary(defaultWorkbasketSummary)
+            .primaryObjRef(defaultObjectReference)
+            .buildAndStore(taskService, "admin");
+    taskService.setTaskRead(task.getId(), true);
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:302]: Does setTaskRead return a Task? If so, you don't need to use "getTask" in the next line

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+    Task task =
+        TaskBuilder.newTask()
+            .classificationSummary(defaultClassificationSummary)
+            .workbasketSummary(defaultWorkbasketSummary)
+            .primaryObjRef(defaultObjectReference)
+            .buildAndStore(taskService, "admin");
+
+    task.setCustomField(CUSTOM_1, "T2100");
+    Task updatedTask = taskService.updateTask(task);
+    updatedTask = taskService.getTask(updatedTask.getId());
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:331]: You don't need to use "getTask", just use "updatedTask"

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 5]
```diff
@@ -0,0 +1,561 @@
...
+
+    assertThat(updatedTask).isNotNull();
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:333]: Please check that the changed attribute is set correctly

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 6]
```diff
@@ -0,0 +1,561 @@
...
+        TaskBuilder.newTask()
+            .classificationSummary(defaultClassificationSummary)
+            .workbasketSummary(defaultWorkbasketSummary)
+            .primaryObjRef(defaultObjectReference)
+            .buildAndStore(taskService, "admin");
+
+    Instant planned = Instant.parse("2024-06-17T23:14:31.0Z");
+    task.setPlanned(planned);
+    Task updatedTask = taskService.updateTask(task);
+    updatedTask = taskService.getTask(updatedTask.getId());
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:349]: You don't need getTask here

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 7]
```diff
@@ -0,0 +1,561 @@
...
+    task.setCustomField(CUSTOM_1, "T2100");
+    Task updatedTask = taskService.updateTask(task);
+    updatedTask = taskService.getTask(updatedTask.getId());
+
+    assertThat(updatedTask).isNotNull();
+  }
+
+  @WithAccessId(user = "user-1-2")
+  @Test
+  void should_SetPlannedCorrectly_When_PlannedSetToALateTime() throws Exception {
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:338]: Name suggestion: "_When_PlannedIsSetExplicitly"

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 8]
```diff
@@ -0,0 +1,561 @@
...
+
+    assertThatThrownBy(() -> taskService.updateTask(task))
+        .isInstanceOf(InvalidArgumentException.class);
+  }
+
+  @WithAccessId(user = "user-1-2")
+  @Test
+  void should_UpdateNoTasks_When_UpdateTasksWithUnmatchedObjectReferenceWasCalled()
+      throws Exception {
+    TaskBuilder.newTask()
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:376]: Why do you need this Task?

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 9]
```diff
@@ -0,0 +1,561 @@
...
+        .workbasketSummary(defaultWorkbasketSummary)
+        .primaryObjRef(objectReference)
+        .buildAndStore(taskService, "admin");
+    TaskBuilder.newTask()
+        .classificationSummary(defaultClassificationSummary)
+        .workbasketSummary(defaultWorkbasketSummary)
+        .primaryObjRef(objectReference)
+        .buildAndStore(taskService, "admin");
+
+    ObjectReferenceImpl por = new ObjectReferenceImpl();
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:420]: Why do you need this? It has the same properties as "objectReference". Could you just reuse "objectReference"?

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 10]
```diff
@@ -0,0 +1,561 @@
...
+      assertThat(task.getCustomField(CUSTOM_10)).isEqualTo("This is modifiedValue 10");
+      assertThat(task.getCustomField(CUSTOM_12)).isEqualTo("This is modifiedValue 12");
+      assertThat(task.getCustomField(CUSTOM_2)).isNotNull();
+      assertThat(task.getCustomField(CUSTOM_16)).isNull();
+    }
+  }
+
+  @WithAccessId(user = "user-1-2")
+  @Test
+  void should_UpdateCallbackInfo_When_RequestedByApi() throws Exception {
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:483]: What do you think about removing "_When_RequestedByApi" from the test name?

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 11]
```diff
@@ -0,0 +1,561 @@
...
+            .buildAndStore(taskService, "admin");
+
+    HashMap<String, String> callbackInfo = new HashMap<>();
+    for (int i = 1; i <= 10; i++) {
+      callbackInfo.put("info_" + i, "Value of info_" + i);
+    }
+    task.setCallbackInfo(callbackInfo);
+    taskService.updateTask(task);
+
+    Task retrievedUpdatedTask = taskService.getTask(task.getId());
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:498]: You don't need to use getTask here

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 12]
```diff
@@ -0,0 +1,561 @@
...
+            .classificationSummary(defaultClassificationSummary)
+            .workbasketSummary(defaultWorkbasketSummary)
+            .primaryObjRef(defaultObjectReference)
+            .buildAndStore(taskService, "admin");
+
+    Instant retrievedTime = Instant.parse("2019-09-13T08:44:17.588Z");
+    task.setReceived(retrievedTime);
+    taskService.updateTask(task);
+
+    Task retrievedUpdatedTask = taskService.getTask(task.getId());
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:516]: You don't need to use getTask here

**Review by ryzheboka — CHANGES_REQUESTED** [review]

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java (hunk 1 above)
Comment (jamesrdi) [reply lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java]: Final has to be used, since distance between declarations and first usage is more than 3 lines (checkstyle violation)
Using .copy() with final declaration causes all task attributes to be set to null, so getTask has to be used

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 13]
```diff
@@ -0,0 +1,582 @@
...
+    task.setBusinessProcessId("MY_PROCESS_ID");
+
+    Task updatedTask = taskService.updateTask(task);
+    updatedTask = taskService.getTask(updatedTask.getId());
+
+    assertThat(updatedTask).isNotNull();
+    assertThat(updatedTask.getCreated()).isNotNull();
+    assertThat(updatedTask.getModified()).isNotNull();
+    assertThat(modifiedOriginal.isAfter(updatedTask.getModified())).isFalse();
+    assertThat(updatedTask.getModified()).isNotEqualTo(updatedTask.getCreated());
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:153]: This check could make the test flaky. If the test runs fast enough, the values of created and modified might be equal. @gitgoodjhe, what do you think about adding sleep after task creation in this test?

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 14]
```diff
@@ -0,0 +1,582 @@
...
+        TaskBuilder.newTask()
+            .classificationSummary(defaultClassificationSummary)
+            .workbasketSummary(defaultWorkbasketSummary)
+            .primaryObjRef(defaultObjectReference)
+            .buildAndStore(taskService, "admin");
+    task.setPrimaryObjRef(
+        taskService.newObjectReference("company", "system", "instance", "value", "type"));
+    ClassificationSummary classificationSummary =
+        defaultTestClassification().buildAndStoreAsSummary(classificationService, "admin");
+    task.setClassificationKey(classificationSummary.getKey());
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:247]: You don't need to change two attributes of Task for this test. What do you think about updating something more simple instead? Like for example the note.

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 15]
```diff
@@ -0,0 +1,582 @@
...
+      throws Exception {
+    Task task =
+        TaskBuilder.newTask()
+            .classificationSummary(defaultClassificationSummary)
+            .workbasketSummary(defaultWorkbasketSummary)
+            .primaryObjRef(defaultObjectReference)
+            .buildAndStore(taskService, "admin");
+    ClassificationSummary classificationSummary =
+        defaultTestClassification().buildAndStoreAsSummary(classificationService, "admin");
+    task.setClassificationKey(classificationSummary.getKey());
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:294]: Please move setting attributes for update (like "setClassificationKey") in the "When" part of the test. Could you do it for other tests too?

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 16]
```diff
@@ -0,0 +1,582 @@
...
+        .isEqualTo(classificationSummary.getKey());
+    assertThat(updatedTask.getClassificationSummary()).isNotEqualTo(defaultClassificationSummary);
+    assertThat(task.getPlanned()).isEqualTo(updatedTask.getPlanned());
+    assertThat(task.getName()).isEqualTo(updatedTask.getName());
+    assertThat(task.getDescription()).isEqualTo(updatedTask.getDescription());
+  }
+
+  @WithAccessId(user = "user-1-2")
+  @Test
+  void should_UpdateReadFlagOfTask_When_SetTaskReadIsCalled() throws Exception {
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:327]: Could you please separate the three test cases from this test into three separate tests? The third test case is not about isRead, but about the TaskNotFoundException thrown during an update

**Review by ryzheboka — CHANGES_REQUESTED** [review]

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 17]
```diff
@@ -0,0 +1,582 @@
...
+        TaskBuilder.newTask()
+            .classificationSummary(defaultClassificationSummary)
+            .workbasketSummary(defaultWorkbasketSummary)
+            .primaryObjRef(defaultObjectReference)
+            .buildAndStore(taskService);
+    task.setBusinessProcessId("MY_PROCESS_ID");
+    task.setParentBusinessProcessId("MY_PARENT_PROCESS_ID");
+
+    Task updatedTask = taskService.updateTask(task);
+    updatedTask = taskService.getTask(updatedTask.getId());
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:127]: You don't need getTask here

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 18]
```diff
@@ -0,0 +1,582 @@
...
+        TaskBuilder.newTask()
+            .classificationSummary(defaultClassificationSummary)
+            .workbasketSummary(defaultWorkbasketSummary)
+            .primaryObjRef(defaultObjectReference)
+            .buildAndStore(taskService);
+    final Instant modifiedOriginal = task.getModified();
+    task.setBusinessProcessId("MY_PROCESS_ID");
+
+    Task updatedTask = taskService.updateTask(task);
+    updatedTask = taskService.getTask(updatedTask.getId());
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:147]: You don't need getTask here

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 19]
```diff
@@ -0,0 +1,582 @@
...
+    Task task =
+        TaskBuilder.newTask()
+            .classificationSummary(defaultClassificationSummary)
+            .workbasketSummary(defaultWorkbasketSummary)
+            .primaryObjRef(defaultObjectReference)
+            .buildAndStore(taskService);
+    task.setPrimaryObjRef(createObjectReference("COMPANY_A", null, null, "VNR", "7654321"));
+
+    Task updatedTask = taskService.updateTask(task);
+    updatedTask = taskService.getTask(updatedTask.getId());
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:191]: You don't need getTask here

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 20]
```diff
@@ -0,0 +1,582 @@
...
+        TaskBuilder.newTask()
+            .classificationSummary(defaultClassificationSummary)
+            .workbasketSummary(defaultWorkbasketSummary)
+            .primaryObjRef(defaultObjectReference)
+            .buildAndStore(taskService, "admin");
+    ClassificationSummary classificationSummary =
+        defaultTestClassification().buildAndStoreAsSummary(classificationService, "admin");
+    task.setClassificationKey(classificationSummary.getKey());
+    Task updatedTask = taskService.updateTask(task);
+    updatedTask = taskService.getTask(updatedTask.getId());
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:314]: Please remove getTask from tests when it's not needed.

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java [hunk 21]
```diff
@@ -0,0 +1,600 @@
...
+    Task updatedTask = taskService.setTaskRead(task.getId(), false);
+
+    assertThat(updatedTask).isNotNull();
+    assertThat(updatedTask.isRead()).isFalse();
+    assertThat(updatedTask.getCreated()).isNotEqualTo(updatedTask.getModified());
+  }
+
+  @WithAccessId(user = "user-1-2")
+  @Test
+  void should_ThrowException_When_SetTaskReadIsCalledWithInvalidTaskId() throws Exception {
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:353]: Please change the name to When_UpdatingTaskWithInvalidId

File: lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java (hunk 13 above)
Comment (gitgoodjhe) [reply lib/taskana-core-test/src/test/java/acceptance/task/update/UpdateTaskAccTest.java:150]: In principle I agree. Are we running with nanosecond prescision or millisecond precision? For nano the question should be out of the window, but even for milli it is highly unlikely that the updateTask runs <1ms.

I think we have similar tests with timestamps that so far never failed

**Review by ryzheboka — APPROVED** [review]
